kernel/kernel-api/src/main/java/io/delta/kernel/internal/commitrange/CommitRangeFactory.java

```diff
@@ -0,0 +1,145 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.commitrange;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static io.delta.kernel.internal.util.Utils.resolvePath;
+
+import io.delta.kernel.CommitRangeBuilder;
+import io.delta.kernel.Snapshot;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.DeltaHistoryManager;
+import io.delta.kernel.internal.DeltaLogActionUtils;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.lang.ListUtils;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.utils.FileStatus;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class CommitRangeFactory {
+
+  private static final Logger logger = LoggerFactory.getLogger(CommitRangeFactory.class);
+
+  private final CommitRangeBuilderImpl.Context ctx;
+  private final Path tablePath;
+  private final Path logPath;
+
+  CommitRangeFactory(Engine engine, CommitRangeBuilderImpl.Context ctx) {
+    this.ctx = ctx;
+    this.tablePath = new Path(resolvePath(engine, ctx.unresolvedPath));
+    this.logPath = new Path(tablePath, "_delta_log");
+  }
+
+  CommitRangeImpl create(Engine engine) {
+    long startVersion = resolveStartVersion(engine);
+    Optional<Long> endVersionOpt = resolveEndVersionIfSpecified(engine);
+    endVersionOpt.ifPresent(
+        endVersion ->
+            checkArgument(
+                startVersion <= endVersion,
+                String.format(
+                    "Resolved startVersion=%d > endVersion=%d", startVersion, endVersion)));
+    logger.info(
+        "{}: Resolved startVersion={} and endVersion={} from startBoundary={} endBoundary={}",
+        tablePath,
+        startVersion,
+        endVersionOpt,
+        ctx.startBoundaryOpt,
+        ctx.endBoundaryOpt);
+    // TODO: for now we just store a list of commit files, this will be updated when we add support
+    //  for ccv2
+    List<FileStatus> deltaFiles =
+        DeltaLogActionUtils.getCommitFilesForVersionRange(
+            engine, tablePath, startVersion, endVersionOpt);
+    // Once we have listed files, we can resolve endVersion=latestVersion for the default case
+    long endVersion =
+        endVersionOpt.orElseGet(
+            () -> FileNames.deltaVersion(ListUtils.getLast(deltaFiles).getPath()));
+    if (!endVersionOpt.isPresent()) {
+      logger.info("{}: Resolved endVersion={} to the latest version", tablePath, endVersion);
+    }
+    return new CommitRangeImpl(
+        tablePath, ctx.startBoundaryOpt, ctx.endBoundaryOpt, startVersion, endVersion, deltaFiles);
+  }
+
+  private SnapshotImpl asSnapshotImpl(Snapshot snapshot) {
```

Comment:
oh -- can't we do this way earlier in the builder during input validation?

Reply:
I'm not sure I want to make the return type of CommitBoundary.getLatestSnapshot -> SnapshotImpl given it's a public API.

I can do the validation earlier but I will still need to cast here.